File: multinomial/c/mixedLogit.c
/*

Naive C Implementation of a Mixed Logit w/ Individual Data 

Utility specification: 
 U_{ijt} = \beta_{0} + \beta_{1}*X_{j} + \beta_{2}*W_{i}*X_{j} +\beta_{3}*S_{i}*X_{j} + \eps_{i,j}

Notes: 
 X is a single product characterstic 
 W is a single consumer specific variable 
 S is a single unobserved taste shock, constant across products and individuals 

After compiling the program (I used standard gcc), the program takes in 3 arguments;
  N_CONS - Number of consumers (100000)
  N_CHOICES - Number of products (10)
  N_SIMS - Number of simulation draws (1000)

e.g, 

mixedLogit 100000 10 1000

*/


#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char **argv){

  int N_CONS = atoi(argv[1]);
  int N_CHOICES = atoi(argv[2]);
  int N_SIMS = atoi(argv[3]);
  printf("\n N_CONS = %d \n N_CHOICES = %d \n N_SIMS = %d \n", N_CONS, N_CHOICES, N_SIMS); 

  /*  FIRST DRAW DGP */

  // Initalize Data Objects  
  double *X = malloc(N_CHOICES*sizeof(double));
  double *W = malloc(N_CONS*sizeof(double));
  double *S = malloc(N_SIMS*sizeof(double));
  int *Y = malloc(N_CONS*sizeof(int));

  // Set Seed 
  srand((unsigned int) time(NULL));

  // Draw U ~ [0,1] for product chars 
  for (int i = 0; i < N_CHOICES; i++){
   X[i] = (float) rand() / (float)(RAND_MAX);
   //printf("%f\n", X[i]);
  }

  // Draw U ~ [0,1] for individual chars 
  for (int i = 0; i < N_CONS; i++){
   W[i] = (float) rand() / (float)(RAND_MAX);
   //printf("%f\n", W[i]);
  }  

  // Draw U ~ [0,1] for random utility draws 
  for (int i = 0; i < N_SIMS; i++){
   S[i] = (float) rand() / (float)(RAND_MAX);
   //printf("%f\n", S[i]);
  }

  // Draw a consumer's choice as a random integer 
  for (int i = 0; i < N_CONS; i++){
   Y[i] = (float) (rand() % N_CHOICES + 1);
   // printf("%f\n", Y[i]);
  }

  // Draw preference parameters from a uniform distribution as well 
  float beta0 = (float) rand() / (float)(RAND_MAX);
  float beta1 = (float) rand() / (float)(RAND_MAX);
  float beta2 = (float) rand() / (float)(RAND_MAX);
  float beta3 = (float) rand() / (float)(RAND_MAX);

  /*  FINDING THE LOG LL */
  clock_t tic = clock();

  // Calculate exponetial of the utils 
  double (*ExpUtils)[N_CHOICES][N_SIMS] = malloc(sizeof(double[N_CHOICES][N_SIMS][N_CONS])); 
  // double ExpUtils[N_CHOICES][N_SIMS][N_CONS];
  for (int i = 0; i < N_CHOICES; i++){
    for (int j = 0; j < N_SIMS; j++){
      for (int k = 0; k < N_CONS; k++){
        ExpUtils[i][j][k] = exp(beta0 + beta1*X[i] + beta2*X[i]*W[k] + beta3*S[j]*X[i]); 
      }
    }
  }
 
  // Caclulate shares given each draw of the sim var
  // double shares[N_CHOICES][N_CONS][N_SIMS];
  double (*shares)[N_CHOICES][N_SIMS] = malloc(sizeof(double[N_CHOICES][N_SIMS][N_CONS]));
  for (int j = 0; j < N_SIMS; j++){
    for (int k = 0; k < N_CONS; k++){
        // calculate the denominator 
        double d = 0; 
        for (int i = 0; i < N_CHOICES; i++){
            d += ExpUtils[i][j][k];
        } 
        // calculate the shares over each simulation draw, given the respective denom 
        for (int i = 0; i < N_CHOICES; i++){
            shares[i][j][k] = ExpUtils[i][j][k]/d;
            // printf("%f\n", ExpUtils[i][0][0]/d);
        } 
    }
  }

  // Integrate over the shares to get CCPs of individuals
  double (*CCP)[N_CHOICES] = malloc(sizeof(double[N_CHOICES][N_CONS])); 
  // double CCP[N_CHOICES][N_CONS];
  for (int i = 0; i < N_CHOICES; i++){
    for (int k = 0; k < N_CONS; k++){
      double ccp_ij = 0;
      for (int j = 0; j < N_SIMS; j++){
        ccp_ij += shares[i][j][k];
      } 
      // *(*(CCP+i)+j) =  ccp_ij/N_SIMS; 
      CCP[i][k] = ccp_ij/N_SIMS; 
      // printf("%f\n", CCP[i][j]);
    }
  }

  // Calculate the LL, loop over consumer's choices and CCP 
  double LL = 0;
  for (int j = 0; j < N_CONS; j++){
    LL -= log((CCP[(Y[j] - 1)][j]));
    // printf("%d\n", Y[j]);
    // printf("%f\n", CCP[(Y[j] - 1)][j]); 
    // printf("%f\n", log(CCP[(Y[j] - 1)][j])); 
  }
  clock_t toc = clock();
  double time_spent = (double)(toc - tic) / CLOCKS_PER_SEC;
  printf(" Time Elasped %f\n", time_spent);

  // free things from memory 
  free(X);
  free(W);
  free(S);
  free(Y);
  free(ExpUtils);
  free(shares);
  free(CCP);

 return 0;
}

File: multinomial/matlab/mixedLogit.m
clear; clc; 

% number of individuals 
N = 100000; 
% number of products 
J = 10; 
% number of simulation draws 
S = 100;

% observable demographics 
W = rand(N,1); 
% product char 
X  = rand(J,1);
% simulation draws, for simulated ln lik 
Draws = randn(1,S);

% beta vector
beta = randn(4,1);

% simulate choices, y (NOTE: THIS IS FAKE AND NOTE DONE GIVEN THE DGP)
y = floor(J*rand(N,1))+1; %agent choices (fake)

%--- SERIAL IMPLEMNTATION ---%

serial_tic = tic; 

% holder for the exp(utils) given a simulation draw
expUtils = ones(J, N, S);
% loop over draws
for i = 1:S
    expUtils(:,:,i) = exp(beta(1) + X*(beta(2) + beta(3)*W' + beta(4)*Draws(i)));
end
% calculate CCPs
CCP = mean(expUtils ./ sum(expUtils, 1), 3);
% calculate lnlik
lnlik_v = ones(N,1); 
for i = 1:N
    lnlik_v(i) = log(CCP(y(i), i));
end
% return negative sum of log like
lnlik = -sum(lnlik_v);

serial_time = toc(serial_tic); 

%--- PARALLEL IMPLEMNTATION ---%

% set the pool size 
P = 20;
% call the pool of size P, takes a hot sec to set up so excluding for runtime measures 
pool = parpool('local', P);

par_tic = tic; 

% holder for the exp(utils) given a simulation draw
expUtils = ones(J, N, S);
% loop over draws
parfor i = 1:S
    expUtils(:,:,i) = exp(beta(1) + X*(beta(2) + beta(3)*W' + beta(4)*Draws(i)));
end
% calculate CCPs
CCP = mean(expUtils ./ sum(expUtils, 1), 3);
% calculate lnlik
lnlik_v = ones(N,1); 
parfor i = 1:N
    lnlik_v(i) = log(CCP(y(i), i));
end
% return negative sum of log like
lnlik = -sum(lnlik_v); 

par_time = toc(par_tic); 

fprintf('Serial Time Elapsed (seconds): %f\n', serial_time)
fprintf('Parallel Time Elapsed (seconds): %f\n', par_time)

% close pool before exiting 
delete(pool);
